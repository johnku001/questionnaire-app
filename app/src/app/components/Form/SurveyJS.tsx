import React, { useState, useEffect } from 'react';
import 'survey-core/defaultV2.min.css';
import { Model } from 'survey-core';
import { Survey } from 'survey-react-ui';

interface SurveyJSProps {
  surveyJson: unknown;
}

function PageSelector({
  pageNo,
  setPageNo,
  pageCount,
}: {
  pageNo: number;
  setPageNo: Function;
  pageCount: number;
}) {
  const getPageSelectorOptions = (): JSX.Element[] => {
    const content: JSX.Element[] = [];
    for (let i = 0; i < pageCount; i++) {
      content.push(
        <option key={i} value={i}>
          Page {i + 1}
        </option>,
      );
    }
    return content;
  };

  return (
    <select
      className="flex-basis-240 sd-input sd-dropdown"
      value={pageNo}
      onChange={evt => {
        setPageNo(JSON.parse(evt.target.value));
      }}
    >
      {getPageSelectorOptions()}
    </select>
  );
}

function ActionButton({
  text,
  onClick,
  canRender,
}: {
  text: string;
  onClick: () => void;
  canRender: boolean;
}) {
  if (!canRender) return null;

  return (
    <button
      className="focus:outline-none focus:ring-2 focus:ring-teal-500 focus:border-teal-500 hover:bg-teal-600 hover:shadow-md transition duration-300 ease-in-out text-base px-8 py-3 rounded-md shadow-md bg-teal-500 text-white"
      onClick={onClick}
    >
      {text}
    </button>
  );
}

export default function SurveyJS({ surveyJson }: SurveyJSProps) {
  const survey = new Model(surveyJson);
  survey.focusFirstQuestion();
  const [pageNo, setPageNo] = useState<number>(survey.currentPageNo);
  const [isRunning, setIsRunning] = useState<boolean>(true);

  useEffect(() => {
    const handlePageChange = (_, options) => {
      setPageNo(options.newCurrentPage.visibleIndex);
    };

    const handleSurveyStart = () => {
      setIsRunning(true);
    };

    const handleSurveyComplete = () => {
      setIsRunning(false);
    };

    survey.onCurrentPageChanged.add(handlePageChange);
    survey.onStarted.add(handleSurveyStart);
    survey.onComplete.add(handleSurveyComplete);

    return () => {
      survey.onCurrentPageChanged.remove(handlePageChange);
      survey.onStarted.remove(handleSurveyStart);
      survey.onComplete.remove(handleSurveyComplete);
    };
  }, [survey]);

  const decrementPageNo = () => setPageNo(pageNo - 1);
  const incrementPageNo = () => setPageNo(pageNo + 1);
  const endSurvey = () => survey.completeLastPage();

  const pageCount = survey.visiblePages.length;

  return (
    <>
      {isRunning && (
        <div className="flex p-6 items-center gap-4 justify-between bg-white">
          <PageSelector
            pageNo={pageNo}
            setPageNo={setPageNo}
            pageCount={pageCount}
          />
          <div className="flex flex-col items-end">
            <div className="flex gap-2">
              <ActionButton
                text="Previous Page"
                onClick={decrementPageNo}
                canRender={pageNo !== 0}
              />
              <ActionButton
                text="Next Page"
                onClick={incrementPageNo}
                canRender={pageNo !== pageCount - 1}
              />
              <ActionButton
                text="Complete"
                onClick={endSurvey}
                canRender={pageNo === pageCount - 1}
              />
            </div>
            <div className="relative w-full">
              <span className="navigation-text absolute right-0 top-2 font-open-sans font-normal text-xs text-gray-400">
                Page {pageNo + 1} of {pageCount}
              </span>
            </div>
          </div>
        </div>
      )}
      <Survey currentPageNo={pageNo} model={survey} />
    </>
  );
}
