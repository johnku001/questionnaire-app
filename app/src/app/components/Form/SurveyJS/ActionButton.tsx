import React from 'react';
import { twMerge } from 'tailwind-merge';

export interface ActionButtonProp extends React.ComponentProps<any> {
  text: string;
  onClick: () => void;
  canRender: boolean;
  className?: string;
  textColor?: string;
  backgroundColor?: string;
}

export function ActionButton({
  text,
  onClick,
  canRender,
  className,
  textColor = '#ffffff',
  backgroundColor = '#246ABC',
}: ActionButtonProp) {
  if (!canRender) return null;

  const buttonStyle = {
    background: backgroundColor,
    color: textColor,
  };

  return (
    <>
      <button
        className={twMerge(
          'focus:outline-none focus:ring-2 focus:opacity-80 hover:opacity-100 hover:shadow-md transition duration-300 ease-in-out opacity-80  text-base px-8 py-3 rounded-md shadow-md  text-white',
          className,
        )}
        onClick={onClick}
        style={buttonStyle}
      >
        {text}
      </button>
    </>
  );
}
