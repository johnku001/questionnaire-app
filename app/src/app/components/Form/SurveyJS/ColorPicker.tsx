import React from 'react';

import { ElementFactory, Question, Serializer } from 'survey-core';
import { SliderPicker, SketchPicker, CompactPicker } from 'react-color';
import {
  SurveyQuestionElementBase,
  ReactQuestionFactory,
} from 'survey-react-ui';
import 'survey-core/defaultV2.min.css';

const CUSTOM_TYPE = 'color-picker';

export class QuestionColorPickerModel extends Question {
  getType() {
    return CUSTOM_TYPE;
  }

  get colorPickerType() {
    return this.getPropertyValue('colorPickerType');
  }
  set colorPickerType(val) {
    this.setPropertyValue('colorPickerType', val);
  }

  get disableAlpha() {
    return this.getPropertyValue('disableAlpha');
  }
  set disableAlpha(val) {
    this.setPropertyValue('disableAlpha', val);
  }
}

// Add question type metadata for further serialization into JSON
Serializer.addClass(
  CUSTOM_TYPE,
  [
    {
      name: 'colorPickerType',
      default: 'Slider',
      choices: ['Slider', 'Sketch', 'Compact'],
    },
    {
      name: 'disableAlpha:boolean',
      dependsOn: 'colorPickerType',
      visibleIf: function (obj: any) {
        return obj.colorPickerType === 'Sketch';
      },
    },
  ],
  function () {
    return new QuestionColorPickerModel('');
  },
  'question',
);

ElementFactory.Instance.registerElement(CUSTOM_TYPE, name => {
  return new QuestionColorPickerModel(name);
});

// A class that renders questions of the new type in the UI
export class SurveyQuestionColorPicker extends SurveyQuestionElementBase {
  constructor(props: any) {
    super(props);
    this.state = { value: this.question.value };
  }
  get question() {
    return this.questionBase;
  }
  get value() {
    return this.question.value;
  }
  get disableAlpha() {
    return this.question.disableAlpha;
  }
  get type() {
    return this.question.colorPickerType;
  }
  handleColorChange = (data: any) => {
    this.question.value = data.hex;
  };
  // Support the read-only and design modes
  get style(): React.CSSProperties | undefined {
    return this.question.getPropertyValue('readOnly') ||
      this.question.isDesignMode
      ? {
          pointerEvents: 'none',
          // Add other styles here as needed
        }
      : undefined;
  }

  renderColor(type: string) {
    switch (type) {
      case 'Slider': {
        return (
          <SliderPicker color={this.value} onChange={this.handleColorChange} />
        );
      }
      case 'Sketch': {
        return (
          <SketchPicker
            color={this.value}
            disableAlpha={this.disableAlpha}
            onChange={this.handleColorChange}
          />
        );
      }
      case 'Compact': {
        return (
          <CompactPicker color={this.value} onChange={this.handleColorChange} />
        );
      }
      default:
        return <div>Unknown type</div>;
    }
  }

  renderElement() {
    return <div style={this.style}>{this.renderColor(this.type)}</div>;
  }
}

// Register `SurveyQuestionColorPicker` as a class that renders `color-picker` questions
ReactQuestionFactory.Instance.registerQuestion(CUSTOM_TYPE, (props: any) => {
  return React.createElement(SurveyQuestionColorPicker, props);
});
