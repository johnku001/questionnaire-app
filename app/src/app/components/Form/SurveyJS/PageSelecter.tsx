import React from 'react';

export interface PageSelectorProps extends React.ComponentProps<any> {
  pageNo: number;
  setPageNo: Function;
  pageCount: number;
}

export function PageSelector({
  pageNo,
  setPageNo,
  pageCount,
}: PageSelectorProps) {
  const getPageSelectorOptions = (): JSX.Element[] => {
    const content: JSX.Element[] = [];
    for (let i = 0; i < pageCount; i++) {
      content.push(
        <option key={i} value={i}>
          Page {i + 1}
        </option>,
      );
    }
    return content;
  };

  return (
    <select
      className="flex-basis-240 sd-input sd-dropdown"
      value={pageNo}
      onChange={evt => {
        setPageNo(JSON.parse(evt.target.value));
      }}
    >
      {getPageSelectorOptions()}
    </select>
  );
}
