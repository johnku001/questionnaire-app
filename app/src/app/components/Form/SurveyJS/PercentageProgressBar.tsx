import * as React from 'react';
import LinearProgress from '@mui/material/LinearProgress';
import Typography from '@mui/material/Typography';
import Box from '@mui/material/Box';

import { Serializer } from 'survey-core';
import { ReactSurveyElement, ReactElementFactory } from 'survey-react-ui';
import 'survey-core/defaultV2.min.css';

Serializer.addProperty('survey', 'progressTitle');

export class PercentageProgressBar extends ReactSurveyElement {
  baseUnit =
    getComputedStyle(document.documentElement).getPropertyValue(
      '--sjs-base-unit',
    ) || '8px';
  calculatedValue = `calc(5 * ${this.baseUnit})`;
  render(): JSX.Element {
    return (
      <Box
        sx={{
          display: 'flex',
          alignItems: 'center',
          paddingX: this.calculatedValue,
        }}
      >
        <Box sx={{ width: '100%', mr: 1 }}>
          <LinearProgress
            variant="determinate"
            value={this.props.model.progressValue}
            {...this.props}
          />
        </Box>
        <Box sx={{ minWidth: 37, textAlign: 'center' }}>
          <Typography variant="body2" color="text.secondary">{`${Math.round(
            this.props.model.progressValue,
          )}%`}</Typography>
        </Box>
      </Box>
    );
  }
}

ReactElementFactory.Instance.registerElement(
  'sv-progressbar-percentage',
  (props: any) => {
    return React.createElement(PercentageProgressBar, props);
  },
);
