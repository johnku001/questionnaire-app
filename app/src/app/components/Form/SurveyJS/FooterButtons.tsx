import React from 'react';
import { Serializer } from 'survey-core';
import { ReactSurveyElement, ReactElementFactory } from 'survey-react-ui';
import { ActionButton } from 'app/components/Form/SurveyJS/ActionButton';
import { themeJson } from 'app/components/Form/theme';

Serializer.addProperty('survey', 'progressTitle');

export class FooterButtons extends ReactSurveyElement {
  render(): JSX.Element {
    return (
      <div className={`flex items-center gap-4 justify-between `}>
        <ActionButton
          text="Start"
          onClick={() => this.props.model.start()}
          canRender={this.props.model.isShowStartingPage}
          textColor={themeJson.cssVariables['--sjs-general-backcolor']}
          backgroundColor={
            themeJson.cssVariables['--sjs-primary-backcolor-dark']
          }
        />
        <ActionButton
          text="Back"
          onClick={() => this.props.model.prevPage()}
          canRender={this.props.model.isShowPrevButton}
          textColor={themeJson.cssVariables['--sjs-primary-backcolor-dark']}
          backgroundColor={themeJson.cssVariables['--sjs-general-backcolor']}
        />

        <ActionButton
          text="Next"
          onClick={() => this.props.model.nextPage()}
          canRender={this.props.model.isShowNextButton}
          className="ml-auto"
          textColor={themeJson.cssVariables['--sjs-primary-backcolor-dark']}
          backgroundColor={themeJson.cssVariables['--sjs-general-backcolor']}
        />
        <ActionButton
          text="Preview"
          onClick={() => this.props.model.showPreview()}
          canRender={
            this.props.model.isLastPage && this.props.model.state !== 'preview'
          }
          className="ml-auto"
          textColor={themeJson.cssVariables['--sjs-primary-backcolor-dark']}
          backgroundColor={themeJson.cssVariables['--sjs-general-backcolor']}
        />
        <ActionButton
          text="Complete"
          onClick={() => this.props.model.completeLastPage()}
          canRender={
            this.props.model.isLastPage && this.props.model.state === 'preview'
          }
          className="ml-auto"
          textColor={themeJson.cssVariables['--sjs-general-backcolor']}
          backgroundColor={
            themeJson.cssVariables['--sjs-primary-backcolor-dark']
          }
        />
      </div>
    );
  }
}

ReactElementFactory.Instance.registerElement('custom-footer', (props: any) => {
  return React.createElement(FooterButtons, props);
});
