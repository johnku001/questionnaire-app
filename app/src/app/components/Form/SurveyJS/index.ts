import './PercentageProgressBar';
import './FooterButtons';
import './ActionButton';
import './ColorPicker';
