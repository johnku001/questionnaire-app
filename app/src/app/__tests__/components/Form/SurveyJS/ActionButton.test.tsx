import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ActionButton } from 'app/components/Form/SurveyJS/ActionButton';
import { createRenderer } from 'react-test-renderer/shallow';

const renderer = createRenderer();

describe('ActionButton', () => {
  it('should render the button and call onClick when clicked', () => {
    const onClickMock = jest.fn();
    const { getByText } = render(
      <ActionButton text="Start" onClick={onClickMock} canRender={true} />,
    );

    const button = getByText('Start');
    fireEvent.click(button);

    expect(button).toBeInTheDocument();
    expect(onClickMock).toHaveBeenCalledTimes(1);
  });

  it('should not render the button when canRender is false', () => {
    const { queryByText } = render(
      <ActionButton text="Start" onClick={() => {}} canRender={false} />,
    );

    const button = queryByText('Start');
    expect(button).toBeNull();
  });

  it('should render and match the snapshot', () => {
    renderer.render(
      <ActionButton text="text" onClick={() => {}} canRender={true} />,
    );
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
