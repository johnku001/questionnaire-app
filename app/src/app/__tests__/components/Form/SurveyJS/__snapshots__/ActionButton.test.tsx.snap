// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton should render and match the snapshot 1`] = `
<React.Fragment>
  <button
    className="focus:outline-none focus:ring-2 focus:opacity-80 hover:opacity-100 hover:shadow-md transition duration-300 ease-in-out opacity-80 text-base px-8 py-3 rounded-md shadow-md text-white"
    onClick={[Function]}
    style={
      Object {
        "background": "#246ABC",
        "color": "#ffffff",
      }
    }
  >
    text
  </button>
</React.Fragment>
`;
